cmd/snaplock/runinhibit/inhibit.go

```diff
@@ -0,0 +1,134 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package runinhibit contains operations for establishing, removing and
+// querying snap run inhibition lock.
+package runinhibit
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+)
+
+// defaultInhibitDir is the directory where inhibition files are stored.
+const defaultInhibitDir = "/var/lib/snapd/inhibit"
+
+// InhibitDir is the directory where inhibition files are stored.
+// This value can be changed by calling dirs.SetRootDir.
+var InhibitDir = defaultInhibitDir
+
+func init() {
+	dirs.AddRootDirCallback(func(root string) {
+		InhibitDir = filepath.Join(root, defaultInhibitDir)
+	})
+}
+
+// Hint is a string representing reason for the inhibition of "snap run".
+type Hint string
+
+const (
+	// HintNotInhibited is used when "snap run" is not inhibited.
+	HintNotInhibited Hint = ""
+	// HintInhibitedForRefresh represents inhibition of a "snap run" while a refresh change is being performed.
+	HintInhibitedForRefresh Hint = "refresh"
+)
+
+func openHintFileLock(snapName string) (*osutil.FileLock, error) {
+	fname := filepath.Join(InhibitDir, snapName+".lock")
+	return osutil.NewFileLockWithMode(fname, 0644)
+}
+
+// LockWithHint sets a persistent "snap run" inhibition lock, for the given snap, with a given hint.
+//
+// The hint cannot be empty. It should be one of the Hint constants defined in
+// this package. With the hint in place "snap run" will not allow the snap to
+// start and will block, presenting a user interface if possible.
```

Comment:
Silly question, what is the behavior of this lock when the system is unexpectedly (manually) rebooted when the lock is in place and not empty? Isn't IsLocked going to give a wrong hint?

Reply:
Clarified on IRC.